File: 프로그래머스/lv3/42884. 단속카메라/단속카메라.py
# n이 10000 이하 이므로 O(n^2)까지 가능

def solution(routes):
    answer = 0
    routes.sort(key=lambda x:x[1]) # 정렬 -> O(nlogn)
    camera = -30001 # 마지막 카메라의 위치
    for route in routes:
        if route[0] > camera: # 카메라의 위치를 벗어나면
            camera = route[1] # 카메라 추가 설치
            answer += 1
    return answer

File: 프로그래머스/3/42884. 단속카메라/단속카메라.java
import java.io.*;
import java.util.*;

class Solution {
    static int s, e;
    
    public int solution(int[][] routes) {
        Arrays.sort(routes, (o1, o2) -> {
            if (o1[0] == o2[0]) return o1[1] - o2[1];
            return o1[0] - o2[0];
        });

        int answer = 1;
        s = routes[0][0];
        e = routes[0][1];
        for (int i = 0; i < routes.length; i++) {
            if (canInclude(routes[i])) {
                reorder(routes[i]);
            } else {
                s = routes[i][0];
                e = routes[i][1];
                answer++;
            }
        }
        return answer;
    }
    
    public void reorder(int[] route) {
        s = Integer.max(s, route[0]);
        e = Integer.min(e, route[1]);
    }
    
    public boolean canInclude(int[] route) {
        if (route[0] > e || route[1] < s) return false;
        return true;
    }
}